File: java/src/me/army8735/jssc/lexer/Token.java
package me.army8735.jssc.lexer;

public class Token {
	public static final int VIRTUAL = -1;
	public static final int OTHER = 0;
	public static final int BLANK = 1;
	public static final int TAB = 2;
	public static final int LINE = 3;
	public static final int NUMBER = 4;
	public static final int ID = 5;
	public static final int COMMENT = 6;
	public static final int STRING = 7;
	public static final int SIGN = 8;
	public static final int REG = 9;
	public static final int KEYWORD = 10;
	public static final int ANNOT = 11;
	public static final int HEAD = 12;
	public static final int TEMPLATE = 13;
	
	private int t;
	private String c;
	private String v;
	
	public Token(int t, String c, String v) {
		this.t = t;
		this.c = c;
		this.v = v;
	}
	
	public int type() {
		return this.t;
	}
	public String content() {
		return this.c;
	}
	public String val() {
		return this.v;
	}
}

File: server/lexer/Token.js
var Class = require('../util/Class'),
	character = require('../util/character'),
	types,
	Token = Class(function(type, content) {
		this.t = type;
		this.c = content;
	}).methods({
		type: function(t) {
			if(!character.isUndefined(t)) {
				this.t = t;
			}
			return this.t;
		},
		content: function() {
			return this.c;
		},
		val: function() {
			return this.v;
		},
		tag: function() {
			return Token.type(this.t);
		},
		toString: function() {
			return this.t + ' ' + this.c;
		}
	}).statics({
		VIRTUAL: -1,
		OTHER: 0,
		BLANK: 1,
		TAB: 2,
		LINE: 3,
		NUMBER: 4,
		ID: 5,
		COMMENT: 6,
		STRING: 7,
		SIGN: 8,
		REG: 9,
		KEYWORD: 10,
		ANNOT: 11,
		HEAD: 12,
		TEMPLATE: 13,
		ENTER: 14,
		type: function(tag) {
			if(character.isUndefined(types)) {
				types = [];
				Object.keys(Token).forEach(function(o) {
					if(typeof Token[o] == 'number') {
						types[Token[o]] = o;
					}
				});
			}
			return types[tag];
		}
	});
module.exports = Token;
